### ISSUE TITLE
MeanStream broken in TF2 (with fixes)

### ISSUE THREAD

**Issue by neel-dey** [voxelmorph/voxelmorph#214]
I believe the migration to TF2 from Keras introduced a few bugs in the `MeanStream` layer used for moving average computation in template construction. They are:
- `cap` is set to trainable, which shouldn't happen.
- `self.mean` and `self.count` do not update and remain at their zero initialization.

The same issues presumably exist in `CovStream` as well, but I haven't tested that layer.

Minimal working example to illustrate the problem (assuming that you have the usual dependencies in `./ext/`):
```python
import sys

sys.path.append('./ext/voxelmorph/')
sys.path.append('./ext/neurite-master/')
sys.path.append('./ext/pynd-lib/')
sys.path.append('./ext/pytools-lib/')

import numpy as np
import tensorflow as tf
import tensorflow.keras.layers as KL
import tensorflow.keras.backend as K

from neurite.tf.layers import MeanStream
from voxelmorph.tf.layers import SpatialTransformer, VecInt

# ----------------------------------------------------------------------------------------------------
# Define and initialize network:

def dummy_net():
    moving_image = tf.keras.layers.Input(shape=[8, 8, 1])
    fixed_image = tf.keras.layers.Input(shape=[8, 8, 1])

    image_input = KL.concatenate([moving_image, fixed_image])

    conv_layer = KL.Conv2D(
        filters=3,
        kernel_size=3,
        padding='same',
        activation='elu',
    )(image_input)

    vel = KL.Conv2D(
        filters=2,
        kernel_size=3,
        padding='same',
        kernel_initializer=tf.keras.initializers.RandomNormal(
            mean=0.0,
            stddev=1e-3,
        ),
        name='vel_field',
    )(conv_layer)

    diff_field = VecInt(method='ss', int_steps=5, name='def_field')(vel)
    diff_field_ms = MeanStream(name='mean_stream', cap=10)(diff_field)
    
    # Transform:
    moved_image = SpatialTransformer()([moving_image, diff_field])
    
    return tf.keras.Model(
        inputs=[moving_image, fixed_image],
        outputs=[moved_image, diff_field_ms],
    )

net = dummy_net()

# ----------------------------------------------------------------------------------------------------
# Set up random moving and fixed images to register and a
# blank deformation field for moving average squared 2-norm loss

dummy_moving = np.random.rand(128, 8, 8, 1)
dummy_fixed = np.random.rand(128, 8, 8, 1)

zero_phi = np.zeros((128, 8, 8, 2))

# ----------------------------------------------------------------------------------------------------
# Compile and inspect model:

mean_stream_loss = lambda _, y_pred: K.mean(K.square(y_pred))

net.compile(
    optimizer='Adam',
    loss=['mse', mean_stream_loss],
    loss_weights=[1., 1.],
)

# Corresponding to cap, mean, and count:
print('number of weights in MeanStream layer: {}'.format(len(net.layers[-1].weights)))

# ----------------------------------------------------------------------------------------------------
# Train model

net.fit(
    [dummy_moving, dummy_fixed],
    [dummy_fixed, zero_phi],
    epochs=2,
    batch_size=4,
)

# ----------------------------------------------------------------------------------------------------
# Inspect MeanStream layer:

# cap should be non trainable, but does get trained:
print('cap weight (should be 10.0): {}'.format(net.layers[-1].weights[0].numpy()))

# mean should be updated, but does not update:
print('mean stream (should be nonzero): {}'.format(net.layers[-1].weights[1].numpy()))

# count should be updated, but does not update:
print('count (should be number of iterations): {}'.format(net.layers[-1].weights[2].numpy()))
```

The output of the last section is (abbreviated for space):
```
cap weight (should be 10.0): 10.071270942687988
mean stream (should be nonzero): [[[0. 0.]
  [0. 0.]
  [0. 0.]
  [0. 0.]
  [0. 0.]
  [0. 0.]
  [0. 0.]
  [0. 0.]]

 ...

 [[0. 0.]
  [0. 0.]
  [0. 0.]
  [0. 0.]
  [0. 0.]
  [0. 0.]
  [0. 0.]
  [0. 0.]]]
count (should be number of iterations): [0.]
```

The fix to the layer is simple enough,
```python
class MeanStream(Layer):
    """ 
    Maintain stream of data mean. 
    cap refers to mainting an approximation of up to that number of subjects -- that is,
    any incoming datapoint will have at least 1/cap weight.
    """

    def __init__(self, cap=100, **kwargs):
        self.cap = tf.Variable(cap, dtype='float32', trainable=False)
        super(MeanStream, self).__init__(**kwargs)

    def build(self, input_shape):
        # Create mean and count
        # These are weights because just maintaining variables don't get saved with the model, and we'd like
        # to have these numbers saved when we save the model.
        # But we need to make sure that the weights are untrainable.
        self.mean = self.add_weight(name='mean', 
                                      shape=input_shape[1:],
                                      initializer='zeros',
                                      trainable=False)
        self.count = self.add_weight(name='count', 
                                      shape=[1],
                                      initializer='zeros',
                                      trainable=False)

        # self.mean = K.zeros(input_shape[1:], name='mean')
        # self.count = K.variable(0.0, name='count')
        super(MeanStream, self).build(input_shape)  # Be sure to call this somewhere!

    def call(self, x):
        # get new mean and count
        this_bs_int = K.shape(x)[0]
        new_mean, new_count = _mean_update(self.mean, self.count, x, self.cap)
        
        # update op
        self.count.assign(new_count)
        self.mean.assign(new_mean)

        # prep for broadcasting :(
        p = tf.concat((K.reshape(this_bs_int, (1,)), K.shape(self.mean)), 0)
        z = tf.ones(p)
        
        # the first few 1000 should not matter that much towards this cost
        return K.minimum(1., new_count/self.cap) * (z * K.expand_dims(new_mean, 0))

    def compute_output_shape(self, input_shape):
        return input_shape
```

And now the output of the final section is as expected:
```
cap weight (should be 10.0): 10.0
mean stream (should be nonzero): [[[-0.07825246 -0.09038504]
  [-0.10282476 -0.15295237]
  [-0.10901316 -0.16708842]
  [-0.10890472 -0.16592586]
  [-0.11188303 -0.16681042]
  [-0.10905293 -0.16895878]
  [-0.09071881 -0.1536277 ]
  [-0.0681491  -0.1155884 ]]

 ...

 [[-0.11265221 -0.08047441]
  [-0.15044352 -0.12022634]
  [-0.14827192 -0.11675054]
  [-0.142835   -0.11020599]
  [-0.14490151 -0.10929355]
  [-0.15004317 -0.11402743]
  [-0.12494531 -0.10127738]
  [-0.08241522 -0.07053082]]]
count (should be number of iterations): [256.]
```

Wasn't sure of whether to open this in the `voxelmorph` repo or `neurite`. Happy to send a PR to `neurite` with the changes :)

**Comment by adalca** [comment]
Hi @neel-dey ,

Thanks! Very interesting, we hadn't noticed this. To be clear, what is the change other than the `trainable=False` for `cap` ? I'm not great at spotting the differences :)

and yup, feel free to do a PR to `neurite` master! Thanks again.

**Comment by neel-dey** [comment]
Hi @adalca,

Ah yes, I should have highlighted the differences, my bad. 

In addition to the change for `cap`, the updates for `self.mean` and `self.count` [in these lines](https://github.com/adalca/neurite/blob/legacy/neuron/layers.py#L1718-L1719) of `MeanStream.call()` have to change to:

```python
self.count.assign(new_count)
self.mean.assign(new_mean)
```

I'll do the `neurite` PR either later tonight or sometime tomorrow. Thanks for the great atlas building framework!

**Comment by adalca** [comment]
Interesting. Do you know what happened in 2.0+? Did `add_update` stop being used? I ask because we have other implementations (like the `CovStream` layer) that use the mechanism to update parameters.

**Comment by neel-dey** [comment]
I'm not too sure of what happened, unfortunately. Going off of [#13941 in `keras`](https://github.com/keras-team/keras/issues/13941) and the [docstring for `add_update`](https://github.com/tensorflow/tensorflow/blob/v2.3.0/tensorflow/python/keras/engine/base_layer.py#L1720) it appears that `add_update` expects an op (or a list of ops) and not a list of (before, after) tuples. When I ran the old keras voxelmorph code back in Feb, `self.mean` and `self.count` updated as expected, so this issue is TF2/tf.keras-specific.

In the context of `MeanStream`, I suppose you could also use `add_update(K.moving_average_update(...))` as in the linked `keras` issue, but there's already an application-specific mechanism for moving averages in the `neurite` layers so I went with the minimal explicit change by using `assign`.

**Comment by neel-dey** [comment]
Closing as addressed by [adalca/neurite#20](https://github.com/adalca/neurite/pull/20). Two additional fixes for `MeanStream` and `CovStream` are made there.
